content/blog/open-source-monitoring-at-the-hpe-customer-technology-center-böblingen.md

```diff
@@ -0,0 +1,377 @@
+            try:
+                # Create a Redfish client object
+                REDFISHOBJ = redfish.redfish_client(base_url=ilo_url, username=login_account, password=login_password)  
+                # Login with the Redfish client
+                REDFISHOBJ.login()
+                s["redfish"]=REDFISHOBJ
+                resource_instances = get_resource_directory(REDFISHOBJ, lfile)
+                for instance in resource_instances:
+                    if '#ComputerSystem.' in instance ['@odata.type']:
+                        s["ComputerSystem"]=instance['@odata.id'] 
+                    if '#Power.' in instance ['@odata.type']:
+                        s["Power"]=instance['@odata.id'] 
+                    if '#Thermal.' in instance ['@odata.type']:
+                        s["Thermal"]=instance['@odata.id']
+                if len(s) > 4:
+                    server_urls[s['ilo']]=s
+                    logwriter(log,s['ilo']+' completed')
+            except Exception as ex:
+                logwriter(log,'Exception - get_server_urls: '+s['ilo'])
+                logwriter(log,str(ex.__context__))
+                if len(s) > 4:
+                    server_urls[s['ilo']]=s
+                pass
+        else:
+            logwriter(log,'Exception - ILO is not reachable: '+s['ilo'])
+    logclose(log)
+    return server_urls
+```
+
+## g﻿et\_server\_data
+
+The get\_server\_data routine is using the obtained server\_urls of one of the monitored server together with the Prometheus user information to retrieve the power and thermal data as well as some general system information data. The captured data is packed into a server\_data directory and returned to the main program.
+
+```python
+def get_server_data( login_account, login_password, server, lfile):
+
+    server_data={}
+    try:
+        REDFISHOBJ = server["redfish"]
+        # System Data
+        server_data['System'] = REDFISHOBJ.get(server["ComputerSystem"]).obj
+        # Power Data
+        uri = REDFISHOBJ.get(server["Power"]).obj.Oem.Hpe['Links']['PowerMeter']['@odata.id']
+        server_data["PowerMeter"] = REDFISHOBJ.get(uri).obj
+        # Temperatures
+        server_data['Temperatures'] = REDFISHOBJ.get(server["Thermal"]).obj["Temperatures"]
+        return server_data;
+    
+    except Exception as ex:
+        sys.stderr.write("ERROR during get_server_date: "+server["url"])
+        if ex.status == 401:
+            REDFISHOBJ = redfish.redfish_client(base_url=server["url"], username=login_account, password=login_password)
+            REDFISHOBJ.login()
+            server["redfish"] = REDFISHOBJ
+        log=logopen(lfile)
+        logwriter(log,'Exception')
+        logwriter(log,str(ex.__context__))
+        logclose(log)        
+        pass
+```
+
+## d﻿isplay\_results
+
+The display\_results routine is taking the server\_data retrieved by the get\_server\_data routine and put it into the corresponding Prometheus node gauges.
```

Comment:
Here, simply say "The display\_results routine takes the server\_data..."

Also, "put" should be "puts" in the second line.